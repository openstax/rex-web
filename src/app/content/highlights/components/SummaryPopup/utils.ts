import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import React from 'react';
import { useSelector } from 'react-redux';
import { systemQueryParameters } from '../../../../navigation/selectors';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const useCreateHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const [link, setLink] = React.useState('');
  const systemQueryParams = useSelector(systemQueryParameters);

  React.useEffect(() => {
    const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
    const data: Record<string, string> = {
      ...systemQueryParams,
      target: JSON.stringify({ id, type: 'highlight' }),
    };
    const target = queryString.stringify(data) + `#${elementId}`;
    setLink(`${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`);
  }, [id, book, elementId, sourceId, systemQueryParams]);

  return link;
};
