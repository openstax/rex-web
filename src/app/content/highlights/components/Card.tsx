import Highlighter, { Highlight } from '@openstax/highlighter';
import { NewHighlightSourceTypeEnum } from '@openstax/highlighter/dist/api';
import { HTMLElement } from '@openstax/types/lib.dom';
import flow from 'lodash/fp/flow';
import React from 'react';
import { connect, useSelector } from 'react-redux';
import styled from 'styled-components';
import { scrollIntoView } from '../../../domUtils';
import { useFocusIn } from '../../../reactUtils';
import { AppState, Dispatch } from '../../../types';
import { highlightStyles } from '../../constants';
import * as selectHighlights from '../../highlights/selectors';
import * as selectSearch from '../../search/selectors';
import * as selectContent from '../../selectors';
import * as contentSelect from '../../selectors';
import { stripIdVersion } from '../../utils/idUtils';
import {
  clearFocusedHighlight,
  createHighlight,
  focusHighlight,
  requestDeleteHighlight,
  setAnnotationChangesPending,
} from '../actions';
import { HighlightData } from '../types';
import { getHighlightLocationFilterForPage } from '../utils';
import { mainCardStyles } from './cardStyles';
import DisplayNote from './DisplayNote';
import EditCard from './EditCard';
import showConfirmation from './utils/showConfirmation';

export interface CardProps {
  page: ReturnType<typeof selectContent['bookAndPage']>['page'];
  book: ReturnType<typeof selectContent['bookAndPage']>['book'];
  container?: HTMLElement;
  isActive: boolean;
  isTocOpen: boolean;
  hasQuery: boolean;
  highlighter: Highlighter;
  highlight: Highlight;
  create: typeof createHighlight;
  focus: typeof focusHighlight;
  remove: typeof requestDeleteHighlight;
  blur: typeof clearFocusedHighlight;
  setAnnotationChangesPending: typeof setAnnotationChangesPending;
  data?: HighlightData;
  className: string;
  zIndex: number;
  shouldFocusCard: boolean;
  topOffset?: number;
  highlightOffsets?: { top: number, bottom: number };
  onHeightChange: (ref: React.RefObject<HTMLElement>) => void;
}

// tslint:disable-next-line:variable-name
const Card = (props: CardProps) => {
  const annotation = props.data && props.data.annotation;
  const element = React.useRef<HTMLElement>(null);
  const [editing, setEditing] = React.useState<boolean>(!annotation);
  const locationFilters = useSelector(selectHighlights.highlightLocationFilters);
  const hasUnsavedHighlight = useSelector(selectHighlights.hasUnsavedHighlight);

  const { isActive, highlight: { id }, focus } = props;

  const focusCard = React.useCallback(async() => {
    if (!isActive && (!hasUnsavedHighlight || await showConfirmation())) {
      focus(id);
    }
  }, [isActive, hasUnsavedHighlight, id, focus]);

  useFocusIn(element, true, focusCard);

  React.useEffect(() => {
    if (!props.isActive) {
      setEditing(false);
    } else {
      const firstElement = props.highlight.elements[0] as HTMLElement;
      const lastElement = props.highlight.elements[props.highlight.elements.length - 1] as HTMLElement;
      const elements = [lastElement];
      if (element.current) {
        elements.push(element.current);
      }
      scrollIntoView(firstElement, elements);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [element, props.isActive]);

  React.useEffect(() => {
    if (annotation) {
      props.highlight.elements.forEach((el) => (el as HTMLElement).classList.add('has-note'));
    } else {
      props.highlight.elements.forEach((el) => (el as HTMLElement).classList.remove('has-note'));
    }
  }, [props.highlight, annotation]);

  React.useEffect(() => {
    if (!annotation && !props.isActive) {
      props.onHeightChange({ current: null });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [annotation, props.isActive]);

  const location = React.useMemo(() => {
    return props.page && getHighlightLocationFilterForPage(locationFilters, props.page);
  }, [locationFilters, props.page]);

  const locationFilterId = location && stripIdVersion(location.id);

  const { page, book } = props;
  if (!props.highlight.range || !page || !book || !locationFilterId) {
    return null;
  }

  const onRemove = () => {
    if (props.data) {
      props.remove(props.data, {
        locationFilterId,
        pageId: page.id,
      });
    }
  };
  const style = highlightStyles.find((search) => props.data && search.label === props.data.color);

  const onCreate = (isDefaultColor: boolean) => {
    props.create({
      ...props.highlight.serialize().getApiPayload(props.highlighter, props.highlight),
      scopeId: book.id,
      sourceId: page.id,
      sourceMetadata: {bookVersion: book.version},
      sourceType: NewHighlightSourceTypeEnum.OpenstaxPage,
    }, {
      isDefaultColor,
      locationFilterId,
      pageId: page.id,
    });
  };

  const commonProps = {
    className: props.className,
    highlight: props.highlight,
    isActive: props.isActive,
    onBlur: props.blur,
    onHeightChange: props.onHeightChange,
    onRemove,
    ref: element,
    shouldFocusCard: props.shouldFocusCard,
  };

  return <div onClick={focusCard} data-testid='card'>
    {
      !editing && style && annotation ? <DisplayNote
        {...commonProps}
        style={style}
        note={annotation}
        focus={props.focus}
        onEdit={() => setEditing(true)}
      /> : <EditCard
        {...commonProps}
        locationFilterId={locationFilterId}
        hasUnsavedHighlight={hasUnsavedHighlight}
        pageId={page.id}
        onCreate={onCreate}
        setAnnotationChangesPending={props.setAnnotationChangesPending}
        onCancel={() => setEditing(false)}
        data={props.data}
      />
    }
  </div>;
};

// tslint:disable-next-line: variable-name
const StyledCard = styled(Card)`
  ${mainCardStyles}
`;

export default connect(
  (state: AppState, ownProps: {highlight: Highlight}) => ({
    ...selectContent.bookAndPage(state),
    data: selectHighlights.highlights(state).find((search) => search.id === ownProps.highlight.id),
    hasQuery: !!selectSearch.query(state),
    isActive: selectHighlights.focused(state) === ownProps.highlight.id,
    isTocOpen: contentSelect.tocOpen(state),
  }),
  (dispatch: Dispatch) => ({
    blur: flow(clearFocusedHighlight, dispatch),
    create: flow(createHighlight, dispatch),
    focus: flow(focusHighlight, dispatch),
    remove: flow(requestDeleteHighlight, dispatch),
    setAnnotationChangesPending: flow(setAnnotationChangesPending, dispatch),
  })
)(StyledCard);
