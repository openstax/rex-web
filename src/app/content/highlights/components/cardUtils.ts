import { Highlight } from '@openstax/highlighter';
import { HighlightColorEnum, HighlightUpdateColorEnum, UpdateHighlightRequest } from '@openstax/highlighter/dist/api';
import { Element, HTMLElement, } from '@openstax/types/lib.dom';
import React from 'react';
import { findElementSelfOrParent } from '../../../domUtils';
import { isHtmlElement } from '../../../guards';
import { assertWindow, remsToPx } from '../../../utils';
import { assertDefined } from '../../../utils/assertions';
import { cardMarginBottom } from '../constants';
import { HighlightData } from '../types';

export const getHighlightOffset = (container: HTMLElement | undefined, highlight: Highlight) => {
  if (!container || !highlight.range || !highlight.range.getBoundingClientRect) {
    return;
  }

  const {top, bottom } = highlight.range.getBoundingClientRect();

  const offsetParent = container.offsetParent && findElementSelfOrParent(container.offsetParent);
  const parentOffset = offsetParent ? offsetParent.offsetTop : 0;
  const scrollOffset = assertWindow().scrollY;

  return {
    bottom: bottom - parentOffset + scrollOffset,
    top: top - parentOffset + scrollOffset,
  };
};

export const getHighlightTopOffset = (container: HTMLElement | undefined, highlight: Highlight): number | undefined => {
  const offset = getHighlightOffset(container, highlight);

  if (offset) {
    return offset.top;
  }
};

export const getHighlightBottomOffset = (
  container: HTMLElement | undefined,
  highlight: Highlight
): number | undefined => {
  const offset = getHighlightOffset(container, highlight);

  if (offset) {
    return offset.bottom;
  }
};

export const generateUpdatePayload = (
  oldData: Pick<HighlightData, 'color' | 'annotation'>,
  update: {color?: HighlightColorEnum, annotation?: string, id: string }
): {
  updatePayload: UpdateHighlightRequest,
  preUpdateData: UpdateHighlightRequest
} => {
  const oldColor = oldData.color as string as HighlightUpdateColorEnum;
  const newColor = update.color !== undefined
    ? update.color as string as HighlightUpdateColorEnum
    : oldColor;

  const updatePayload = {
    highlight: {
      annotation: update.annotation !== undefined ? update.annotation : oldData.annotation,
      color: newColor,
    },
    id: update.id,
  };

  const preUpdateData = {
    highlight: {
      annotation: oldData.annotation,
      color: oldColor,
    },
    id: update.id,
  };

  return {updatePayload, preUpdateData};
};

/**
 * Calculate positions for the cards according to the location of associated highlights in the document.
 * Positions will be adjusted to the presence of the other cards in cases like multiple highlights in the same line.
 */
const updateStackedCardsPositions = (
  highlightsElements: Highlight[],
  heights: Map<string, number>,
  getHighlightPosition: (highlight: Highlight) => { top: number, bottom: number },
  initialPositions?: Map<string, number>,
  addAditionalMarginForTheFirstCard = false,
  lastVisibleCardPosition = 0,
  lastVisibleCardHeight = 0
) => {
  const positions = initialPositions ? initialPositions : new Map<string, number>();

  for (const [index, highlight] of highlightsElements.entries()) {
    const topOffset = getHighlightPosition(highlight).top;

    const marginToAdd = index > 0 || addAditionalMarginForTheFirstCard ? remsToPx(cardMarginBottom) : 0;
    const lastVisibleCardBottom = lastVisibleCardPosition + lastVisibleCardHeight;
    const stackedTopOffset = Math.max(topOffset, lastVisibleCardBottom + marginToAdd);

    if (heights.get(highlight.id)) {
      lastVisibleCardPosition = stackedTopOffset;
      lastVisibleCardHeight = heights.get(highlight.id)!;
    }

    positions.set(highlight.id, stackedTopOffset);
  }

  return positions;
};

/**
 * Calculate how much we have to move cards to adjust their offset so we can align a card for @param highlight
 * with the corresponding highlight in the document.
 */
const getOffsetToAdjustForHighlightPosition = (
  highlight: Highlight | undefined,
  cardsPositions: Map<string, number>,
  getHighlightPosition: (highlight: Highlight) => { top: number, bottom: number }
) => {
  const position = highlight
    ? assertDefined(cardsPositions.get(highlight.id), 'internal function requested postion of unknown highlight')
    : 0;

  const topOffsetFocused = highlight && position
    ? getHighlightPosition(highlight).top
    : 0;

  return position - topOffsetFocused;
};

/**
 * Calculate positions of all cards for @param highlights
 * If @param focusedHighlight is specified then change position of the related card to align it
 * with the corresponding highlight in the document and recalculate positions of all other cards that
 * may be affected by this operation.
 */
export const updateCardsPositions = (
  focusedHighlight: Highlight | undefined,
  highlights: Highlight[],
  cardsHeights: Map<string, number>,
  getHighlightPosition: (highlight: Highlight) => { top: number, bottom: number }
) => {
  const cardsPositions = updateStackedCardsPositions(highlights, cardsHeights, getHighlightPosition);

  const offsetToAdjust = getOffsetToAdjustForHighlightPosition(focusedHighlight, cardsPositions, getHighlightPosition);

  if (!focusedHighlight || offsetToAdjust === 0) { return cardsPositions; }

  const focusedHighlightIndex = highlights.findIndex((highlight) => highlight.id === focusedHighlight.id);

  // Start processing the first part of highlights

  const highlightsUpToFocused = highlights.slice(0, focusedHighlightIndex + 1);

  // Reverse an array to start from the focused highlight card, adjust its position and continue adjusting cards above
  const reversedHighlightsUpToFocused = highlightsUpToFocused.reverse();

  for (const [index, highlight] of reversedHighlightsUpToFocused.entries()) {
    const highlightCardTopOffset = cardsPositions.get(highlight.id) as number;
    const adjustedHighlightPosition = highlightCardTopOffset - offsetToAdjust;

    // Adjust position of the card to the calculated offset
    cardsPositions.set(highlight.id, adjustedHighlightPosition);

    // Check if we have to adjust position of a card above and if not then simply break
    const nextHighlight = reversedHighlightsUpToFocused[index + 1] as Highlight | undefined;
    if (nextHighlight) {
      const nextHighlightCardTopOffset = cardsPositions.get(nextHighlight.id) as number;
      const nextHighlightCardHeight = cardsHeights.get(nextHighlight.id) as number;
      const nextHighlightCardBottomOffset = nextHighlightCardTopOffset + nextHighlightCardHeight;

      if (nextHighlightCardBottomOffset <= adjustedHighlightPosition) {
        break;
      }
    }
  }

  // Start processing the second part of highlights

  // Adjusting the position of focused highlight might cause some cards to go up which means that maybe cards
  // that were previously pushed down now have space to align with their highlights.

  const highlightsAfterFocused = highlights.slice(focusedHighlightIndex + 1);

  return updateStackedCardsPositions(
    highlightsAfterFocused,
    cardsHeights,
    getHighlightPosition,
    cardsPositions,
    true,
    cardsPositions.get(focusedHighlight.id) as number,
    cardsHeights.get(focusedHighlight.id) as number
  );
};

export const noopKeyCombinationHandler = (
  allowedContainers: Array<HTMLElement | React.RefObject<HTMLElement>>
) => (activeElement: Element): boolean => {
  const activeElementIsOutsideOfTheAllowedContainer = allowedContainers.every((container) => {
    return isHtmlElement(container)
      ? !container.contains(activeElement)
      : !container.current || !container.current.contains(activeElement);
  });
  if (activeElementIsOutsideOfTheAllowedContainer) { return true; }
  if (activeElement.nodeName === 'TEXTAREA') { return true; }
  return false;
};
