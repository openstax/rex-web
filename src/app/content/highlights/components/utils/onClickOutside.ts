import { Element, EventListenerOptions, HTMLElement, MouseEvent } from '@openstax/types/lib.dom';
import curry from 'lodash/fp/curry';
import React from 'react';
import { isHtmlElement } from '../../../../guards';
import { assertDocument, assertWindow } from '../../../../utils';

const targetBelongsToElement = curry((event: MouseEvent, el: React.RefObject<HTMLElement> | HTMLElement): boolean =>
  isHtmlElement(el)
    ? el.contains(event.target as Element)
    : el.current
      ? el.current.contains(event.target as Element)
      : false
);

const onClickOutside = (
  element: React.RefObject<HTMLElement> | Array<React.RefObject<HTMLElement> | HTMLElement>,
  isEnabled: boolean,
  cb: (ev: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => () => {
  if (typeof document === 'undefined') {
    return () => null;
  }

  const ifOutside = (e: MouseEvent) => {
    if (!(e.target instanceof assertWindow().Element)) {
      return;
    } else if (Array.isArray(element) && element.some(targetBelongsToElement(e))) {
      return;
    } else if (!Array.isArray(element) && targetBelongsToElement(e, element)) {
      return;
    }
    cb(e);
  };

  if (isEnabled) {
    document.addEventListener('click', ifOutside, eventOptions);
  }

  return () => assertDocument().removeEventListener('click', ifOutside, eventOptions);
};

export const useOnClickOutside = (
  element: React.RefObject<HTMLElement> | Array<React.RefObject<HTMLElement> | HTMLElement>,
  isEnabled: boolean,
  cb: (e: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  React.useEffect(onClickOutside(element, isEnabled, cb, eventOptions), [element, cb, isEnabled]);
};

export default onClickOutside;

const isRefWithHtmlElement = (el: any): el is React.RefObject<HTMLElement> => {
  return el instanceof Object && isHtmlElement(el.current);
};

export const isElementForOnClickOutside = (el: any): el is HTMLElement | React.RefObject<HTMLElement> => {
  return isHtmlElement(el) || isRefWithHtmlElement(el);
};
