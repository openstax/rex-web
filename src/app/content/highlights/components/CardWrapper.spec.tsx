import { Highlight } from '@openstax/highlighter';
import { Document, HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestStore from '../../../../test/createTestStore';
import createMockHighlight from '../../../../test/mocks/highlight';
import * as domUtils from '../../../domUtils';
import { Store } from '../../../types';
import { assertDocument, remsToPx } from '../../../utils';
import { assertWindow } from '../../../utils/browser-assertions';
import { focusHighlight } from '../actions';
import { cardMarginBottom, highlightKeyCombination } from '../constants';
import Card from './Card';
import * as cardUtils from './cardUtils';
import CardWrapper from './CardWrapper';

const dispatchKeyDownEvent = (
  window: Window,
  element: Document | HTMLElement,
  key: string,
  target?: HTMLElement
) => {
  const keyboardEvent = new KeyboardEvent('keydown', { bubbles: true, cancelable: true, key, view: window });
  if (target) {
    Object.defineProperty(keyboardEvent, 'target', { value: target });
  }
  element.dispatchEvent(keyboardEvent);
};

const dispatchFocusOutEvent = (
  window: Window,
  element: Document | HTMLElement,
  relatedTarget: HTMLElement
) => {
  const focusEvent = window.document.createEvent('FocusEvent');
  focusEvent.initEvent('focusout', true, true);
  Object.defineProperty(focusEvent, 'relatedTarget', { value: relatedTarget });
  element.dispatchEvent(focusEvent);
};

jest.mock('./Card', () => (props: any) => <span data-mock-card {...props} />);

jest.mock('./cardUtils', () => ({
  ...jest.requireActual('./cardUtils'),
  getHighlightOffset: jest.fn(() => ({ top: 100, bottom: 100 })),
}));

describe('CardWrapper', () => {
  let store: Store;
  let container: HTMLElement;

  beforeEach(() => {
    store = createTestStore();
    container = assertDocument().createElement('div');
    assertDocument().body.appendChild(container);
  });

  afterEach(() => {
    container.remove();
  });

  it('matches snapshot', () => {
    const component = renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[createMockHighlight('id1')]} />
    </Provider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot when there is no highlights', () => {
    const component = renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[]} />
    </Provider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders cards', () => {
    const component = renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[createMockHighlight(), createMockHighlight()]} />
    </Provider>);

    expect(component.root.findAllByType(Card).length).toBe(2);
  });

  it('scrolls to card when focused', () => {
    const scrollIntoView = jest.spyOn(domUtils, 'scrollIntoView');
    scrollIntoView.mockImplementation(() => null);

    const highlight = {
      ...createMockHighlight(),
      elements: ['something'],
    };

    renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[highlight]} />
    </Provider>);

    renderer.act(() => {
      store.dispatch(focusHighlight(highlight.id));
    });

    expect(scrollIntoView).toHaveBeenCalled();
    scrollIntoView.mockClear();
  });

  it('do not scroll to card when focused if it does not have elements', () => {
    const scrollIntoView = jest.spyOn(domUtils, 'scrollIntoView');
    scrollIntoView.mockImplementation(() => null);

    const highlight = {
      ...createMockHighlight(),
      elements: [],
    };

    renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[highlight]} />
    </Provider>);

    renderer.act(() => {
      store.dispatch(focusHighlight(highlight.id));
    });

    expect(scrollIntoView).not.toHaveBeenCalled();
  });

  it('do not scroll to focused card after rerender if focused highlight is the same', () => {
    const scrollIntoView = jest.spyOn(domUtils, 'scrollIntoView');
    scrollIntoView.mockImplementation(() => null);

    const highlight = {
      ...createMockHighlight(),
      elements: ['something'],
    };
    const highlight2 = {
      ...createMockHighlight(),
      elements: ['else'],
    };
    const highlight3 = {
      ...createMockHighlight(),
      elements: ['woops'],
    };

    const component = renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[highlight, highlight2]} />
    </Provider>);

    renderer.act(() => {
      store.dispatch(focusHighlight(highlight.id));
    });

    expect(scrollIntoView).toHaveBeenCalled();
    scrollIntoView.mockClear();

    component.update(<Provider store={store}>
      <CardWrapper container={container} highlights={[highlight, highlight2, highlight3]} />
    </Provider>);

    // make sure that useEffect is called
    // tslint:disable-next-line: no-empty
    renderer.act(() => {});

    expect(scrollIntoView).not.toHaveBeenCalled();
  });

  it(`handles card's height changes`, () => {
    const component = renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[createMockHighlight(), createMockHighlight()]} />
    </Provider>);

    const [card1, card2] = component.root.findAllByType(Card);
    expect(card1.props.topOffset).toEqual(undefined);
    expect(card2.props.topOffset).toEqual(undefined);

    // Update state with a height
    renderer.act(() => {
      card1.props.onHeightChange({ current: { offsetHeight: 100 } });
      card2.props.onHeightChange({ current: { offsetHeight: 100 } });
    });
    // We are starting at 100 because of getHighlightTopOffset mock
    expect(card1.props.topOffset).toEqual(100);
    expect(card2.props.topOffset).toEqual(100 + 100 + remsToPx(cardMarginBottom));

    // Noops when height is the same
    renderer.act(() => {
      card1.props.onHeightChange({ current: { offsetHeight: 100 } });
    });
    expect(card1.props.topOffset).toEqual(100);
    expect(card2.props.topOffset).toEqual(100 + 100 + remsToPx(cardMarginBottom));

    // Handle null value
    renderer.act(() => {
      card1.props.onHeightChange({ current: { offsetHeight: null } });
    });
    // First card have null height so secondcard starts at the highlight top offset
    expect(card1.props.topOffset).toEqual(100);
    expect(card2.props.topOffset).toEqual(100);

    expect(() => component.root.findAllByType(Card)).not.toThrow();
  });

  it(`update only these cards that needs it when focusing highlight`, () => {
    const highlights = [createMockHighlight(), createMockHighlight(), createMockHighlight(), createMockHighlight()];
    const highlightsPositionsInDocument = [0, 100, 120, 300];

    jest.spyOn(cardUtils, 'getHighlightOffset')
      .mockImplementation((_container: any, highlight: Highlight) => {
        const top = highlightsPositionsInDocument[highlights.findIndex((search) => search.id === highlight.id)];
        return {
          bottom: top + 20,
          top,
        };
      });

    const component = renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={highlights} />
    </Provider>);

    const [card1, card2, card3, card4] = component.root.findAllByType(Card);
    expect(card1.props.topOffset).toEqual(undefined);
    expect(card2.props.topOffset).toEqual(undefined);
    expect(card3.props.topOffset).toEqual(undefined);
    expect(card3.props.topOffset).toEqual(undefined);

    // Update state with a height
    renderer.act(() => {
      card1.props.onHeightChange({ current: { offsetHeight: 50 } });
      card2.props.onHeightChange({ current: { offsetHeight: 50 } });
      card3.props.onHeightChange({ current: { offsetHeight: 50 } });
      card4.props.onHeightChange({ current: { offsetHeight: 50 } });
    });
    expect(card1.props.topOffset).toEqual(0); // first card have only 50px height + 20px margin bottom
    expect(card2.props.topOffset).toEqual(100); // so the second cards starts and the highlight level which is 100
    // highlight for the card3 starts at 120px which is too close to the card2
    // calculated offset will be equal to card2 offset + card2 height + card2 margin bottom = 170px
    expect(card3.props.topOffset).toEqual(170);
    expect(card4.props.topOffset).toEqual(300); // card4 will start at the corresponding highlight level

    // Move card2 and card3 to the top when the card3 is focused - card1 and card4 are not affected
    renderer.act(() => {
      store.dispatch(focusHighlight(card3.props.highlight.id));
    });
    // cards were moved by 70px up (50 is height of the focused card and 20 px is margin bottom)
    expect(card1.props.topOffset).toEqual(0);
    expect(card2.props.topOffset).toEqual(50);
    expect(card3.props.topOffset).toEqual(120);
    expect(card4.props.topOffset).toEqual(300);

    // focusing card1 is not changing position of any cards
    renderer.act(() => {
      store.dispatch(focusHighlight(card1.props.highlight.id));
    });
    // cards were moved by 70px up (50 is height of the focused card and 20 px is margin bottom)
    expect(card1.props.topOffset).toEqual(0);
    expect(card2.props.topOffset).toEqual(100);
    expect(card3.props.topOffset).toEqual(170);
    expect(card4.props.topOffset).toEqual(300);

    // focusing card4 is not changing position of any cards
    renderer.act(() => {
      store.dispatch(focusHighlight(card1.props.highlight.id));
    });
    // cards were moved by 70px up (50 is height of the focused card and 20 px is margin bottom)
    expect(card1.props.topOffset).toEqual(0);
    expect(card2.props.topOffset).toEqual(100);
    expect(card3.props.topOffset).toEqual(170);
    expect(card4.props.topOffset).toEqual(300);

    expect(() => component.root.findAllByType(Card)).not.toThrow();
  });

  it('does not throw on unmount when highlight is focused', () => {
    const highlight = createMockHighlight();
    store.dispatch(focusHighlight(highlight.id));

    const component = renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[highlight]} />
    </Provider>);

    expect(() => component.unmount()).not.toThrow();
  });

  it('handles useKeyCombination - focus highlight in the content', () => {
    const window = assertWindow();
    const document = assertDocument();
    const highlight = createMockHighlight();
    const highlightElement = document.createElement('span');
    container.appendChild(highlightElement);

    const cardWrapperElement = document.createElement('div');
    const cardElement = document.createElement('div');
    cardWrapperElement.append(cardElement);

    store.dispatch(focusHighlight(highlight.id));

    renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[highlight]} />
    </Provider>, { createNodeMock: () => cardWrapperElement });

    renderer.act(() => undefined);

    expect(store.getState().content.highlights.currentPage.focused).toEqual(highlight.id);

    dispatchKeyDownEvent(window, document, highlightKeyCombination.key!, cardElement);

    expect(highlight.focus).toHaveBeenCalled();
  });

  it('handles useKeyCombination - focus card and then handles useFocusLost - unfocus card', () => {
    const window = assertWindow();
    const document = assertDocument();
    const highlight = createMockHighlight();
    const highlightElement = document.createElement('span');
    container.appendChild(highlightElement);

    const cardWrapperElement = document.createElement('div');

    store.dispatch(focusHighlight(highlight.id));

    const component = renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[highlight]} />
    </Provider>, { createNodeMock: () => cardWrapperElement });

    renderer.act(() => {
      const card = component.root.findByType(Card);
      expect(card.props.shouldFocusCard).toEqual(false);
    });

    dispatchKeyDownEvent(window, document, highlightKeyCombination.key!, highlightElement);

    renderer.act(() => {
      const card = component.root.findByType(Card);
      expect(card.props.shouldFocusCard).toEqual(true);
    });

    expect(highlight.focus).not.toHaveBeenCalled();

    const elementOutside = document.createElement('span');

    dispatchFocusOutEvent(window, cardWrapperElement, elementOutside);

    renderer.act(() => {
      const card = component.root.findByType(Card);
      expect(card.props.shouldFocusCard).toEqual(false);
    });
  });

  it(
    'handles useKeyCombination - noop if trigerred in element that we dont support '
    + 'or with another key combination',
    () => {
    const window = assertWindow();
    const document = assertDocument();
    const highlight = createMockHighlight();
    const highlightElement = document.createElement('span');
    container.appendChild(highlightElement);

    const textarea = document.createElement('textarea');
    container.appendChild(textarea);

    const elementInsideContainer = document.createElement('div');
    container.appendChild(elementInsideContainer);

    const elementOutsideOfTheContainer = document.createElement('div');
    document.body.appendChild(elementOutsideOfTheContainer);

    const cardWrapperElement = document.createElement('div');

    store.dispatch(focusHighlight(highlight.id));

    const component = renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[highlight]} />
    </Provider>, { createNodeMock: () => cardWrapperElement });

    renderer.act(() => {
      const card = component.root.findByType(Card);
      expect(card.props.shouldFocusCard).toEqual(false);
    });

    dispatchKeyDownEvent(window, document, highlightKeyCombination.key!, textarea);

    dispatchKeyDownEvent(window, document, highlightKeyCombination.key!, elementOutsideOfTheContainer);

    dispatchKeyDownEvent(window, document, 'anotherkeythatwedontsupport', elementInsideContainer);

    renderer.act(() => {
      const card = component.root.findByType(Card);
      expect(card.props.shouldFocusCard).toEqual(false);
    });

    expect(highlight.focus).not.toHaveBeenCalled();
  });

  it('handles useKeyCombination - noop if focusedHighlight is undefined', () => {
    const window = assertWindow();
    const document = assertDocument();
    const highlight = createMockHighlight();
    const highlightElement = document.createElement('span');
    container.appendChild(highlightElement);

    const cardWrapperElement = document.createElement('div');
    const cardElement = document.createElement('div');
    cardWrapperElement.append(cardElement);

    renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[highlight]} />
    </Provider>, { createNodeMock: () => cardWrapperElement });

    renderer.act(() => undefined);

    expect(store.getState().content.highlights.currentPage.focused).toEqual(undefined);

    dispatchKeyDownEvent(window, document, highlightKeyCombination.key!, cardElement);

    expect(highlight.focus).not.toHaveBeenCalled();
  });

  it('handles useKeyCombination - noop if event.target is undefined', () => {
    const window = assertWindow();
    const document = assertDocument();
    const highlight = createMockHighlight();
    const highlightElement = document.createElement('span');
    container.appendChild(highlightElement);

    const cardWrapperElement = document.createElement('div');
    const cardElement = document.createElement('div');
    cardWrapperElement.append(cardElement);

    store.dispatch(focusHighlight(highlight.id));

    renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[highlight]} />
    </Provider>, { createNodeMock: () => cardWrapperElement });

    renderer.act(() => undefined);

    expect(store.getState().content.highlights.currentPage.focused).toEqual(highlight.id);

    dispatchKeyDownEvent(window, document, highlightKeyCombination.key!, undefined);

    expect(highlight.focus).not.toHaveBeenCalled();
  });

  it('handles useKeyCombination - noop if element.current is undefined', () => {
    const window = assertWindow();
    const document = assertDocument();
    const highlight = createMockHighlight();
    const highlightElement = document.createElement('span');
    container.appendChild(highlightElement);

    const cardWrapperElement = document.createElement('div');
    const cardElement = document.createElement('div');
    cardWrapperElement.append(cardElement);

    store.dispatch(focusHighlight(highlight.id));

    renderer.create(<Provider store={store}>
      <CardWrapper container={container} highlights={[highlight]} />
    </Provider>, { createNodeMock: () => undefined });

    renderer.act(() => undefined);

    expect(store.getState().content.highlights.currentPage.focused).toEqual(highlight.id);

    dispatchKeyDownEvent(window, document, highlightKeyCombination.key!, cardElement);

    expect(highlight.focus).not.toHaveBeenCalled();
  });
});
