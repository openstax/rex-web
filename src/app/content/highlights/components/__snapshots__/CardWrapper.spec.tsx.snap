// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardWrapper matches snapshot 1`] = `
<div
  className="sc-bdVaJa jZCMyL"
>
  <span
    container={<div />}
    data-mock-card={true}
    highlight={
      Object {
        "addFocusedStyles": [MockFunction],
        "elements": Array [],
        "focus": [MockFunction],
        "getStyle": [MockFunction],
        "id": "id1",
        "isAttached": [MockFunction],
        "range": Object {
          "commonAncestorContainer": null,
          "getBoundingClientRect": [MockFunction],
        },
        "serialize": [Function],
        "setStyle": [MockFunction],
        "sourceId": "testbook1-testpage1-uuid",
      }
    }
    onHeightChange={[Function]}
    shouldFocusCard={false}
    zIndex={1}
  />
</div>
`;

exports[`CardWrapper matches snapshot when there is no highlights 1`] = `null`;
